bufferserver/src/main/java/com/datatorrent/bufferserver/internal/PhysicalNode.java

```diff
@@ -71,36 +71,29 @@ public long getUptime()
    * @param d
    * @throws InterruptedException
    */
-  private SerializedData blocker;
-
   public boolean send(SerializedData d)
   {
-    if (d.offset == d.dataOffset) {
-      if (client.write(d.buffer, d.offset, d.length)) {
-        return true;
-      }
-    } else {
-      if (client.send(d.buffer, d.offset, d.length)) {
-        return true;
-      }
+    if (client.send(d.buffer, d.dataOffset, d.length - (d.dataOffset - d.offset))) {
+      return true;
+    }
+    if (blocker == null) {
+      blocker = d;
+    } else if (blocker != d) {
+      throw new IllegalStateException(String.format("Can't send data %s while blocker %s is pending on %s", d, blocker, this));
     }
-
-    blocker = d;
     return false;
   }
 
   public boolean unblock()
   {
-    if (blocker == null) {
-      return true;
-    }
-
-    if (send(blocker)) {
-      blocker = null;
-      return true;
+    if (blocker != null) {
+      if (client.send(blocker.buffer, blocker.dataOffset, blocker.length - (blocker.dataOffset - blocker.offset))) {
```

Comment:
Refactored to return true when either blocker is null or client.send() returns true. It looks more readable to me, I can undo the refactoring.